rfcs/20200106-prelu/README.md

```diff
@@ -0,0 +1,124 @@
+
+```
+n = 256, h = 7, w = 7 c = 128
+src_md = {dims={256, 128, 7, 7}, f32, tag::nhwc};
+```
+
+| Case                              | Primitive Descriptor Initialization
+| :--                               | :--
+| Whole-tensor (TensorFlow variant) | `prelu_forward::desc(src_md, weights_md={{256, 128, 7, 7}, tag::nhwc}})`
+| Channel-wise                      | `prelu_forward::desc(src_md, weights_md={{1, 128, 1, 1}, tag::nhwc}})`
+| Channel-shared                    | `prelu_forward::desc(src_md, weights_md={{1, 1, 1, 1}, tag::nhwc}})`
+| Shared axes (n and c shared)      | `prelu_forward::desc(src_md, weights_md={{1, 1, 7, 7}, tag::nhwc}})`
+
+The library will recommend user to allow PReLU primitive to choose the
+appropriate weights memory format by passing weights_md with `format_tag::any`.
+This might be important for the whole-tensor case, where ideally the weights
+should repeat the memory format of the data tensor. However, given that the
+most typical case will be channel-wise (`c` elements) or channel-shared (1
+element) weights, the primitive should be optimized for the plain format that
+user has. Data type will be preserved.
+
+Performance friendly examples:
+- `prelu_forward::desc(src_md, weights_md={{n, c, h, w}, f32, tag::any}})`
+- `prelu_forward::desc(src_md, weights_md={{1, c, 1, 1}, f32, tag::any}})`
+- `prelu_forward::desc(src_md, weights_md={{1, c, 1, 1}, f32, tag::abcd}})`
+- `prelu_forward::desc(src_md, weights_md={{1, 1, 1, 1}, f32, tag::abcd}})`
+
+Not recommended (for **full** `(n, c, h, w)` weights tensor the memory format
+is forced):
+- `prelu_forward::desc(src_md, weights_md={{n, c, h, w}, tag::abcd}})`
+
+
+### Option B - extend existing eltwise implementation
+
+### Recommendation
+
+I would recommend Option A. In case of calculation gradients for alpha we must
+pass memory descriptors for alpha and alpha_grad. Implementation in existing
+eltwise kernel would require breaking the ABI. What is more now in eltwise we
+pass float as alpha parameter. If we added `DNNL_ARG_WEIGHTS` (alpha) required
+for PReLU and alpha required for LeakyReLU remained it would be ambiguous for
+the user.
+
+## Open questions
+
+1. TensorFlow gives user greater flexibility than channel wise/shared,
+   by shared_axes parameter. Should we support that as well?
+
+   API wise yes, implementation of full functionality can be spread over time
+   according to the requests and priorities.
+
+2. Should the primitive have only cumulative `backward` pass (that includes
+   backward by data and by weights), or should it support those steps
+   separately?
+   - Separate pass might be required, if at some point the alpha is no more
+     treated as learnable parameter.
+   - It is not clear if cumulative backward pass brings any value, though not
+     providing it is strange, as other primitives (batch and layer
+     normalization) have it.
+
+   Current suggestion: support `backward = backward_data + backward_weights`,
+   and `backward_data`.
```

Comment:
One more thing that came to my mind is specification for the backward pass.
Would you mind to check with FWK folks if they are fine with the suggestion?

Reply:
Sure I will ask them.

Reply:
_For PyTorch side, it has just have one path(weight is always a learnable parameter): cumulative backward pass, see prelu_backward_cpu, so the suggestion is fine for me._